flink-runtime/src/main/java/org/apache/flink/runtime/operators/hash/ReduceHashTable.java

```diff
@@ -0,0 +1,1014 @@
+     */
+	@Override
+	public void insertOrReplaceRecord(T record) throws IOException {
+		if (closed) {
+			return;
+		}
+
+		T match = prober.getMatchFor(record, reuse);
+		if (match == null) {
+			prober.insertAfterNoMatch(record);
+		} else {
+			prober.updateMatch(record);
+		}
+	}
+
+	/**
+	 * Inserts the given record into the hash table.
+	 * Note: this method doesn't care about whether a record with the same key is already present.
+	 * @param record The record to insert.
+	 * @throws IOException (EOFException specifically, if memory ran out)
+     */
+	@Override
+	public void insert(T record) throws IOException {
+		if (closed) {
+			return;
+		}
+
+		final int hashCode = MathUtils.hash(buildSideComparator.hash(record));
+		final int bucket = hashCode & numBucketsMask;
+		final int bucketSegmentIndex = bucket >>> numBucketsPerSegmentBits; // which segment contains the bucket
+		final MemorySegment bucketSegment = bucketSegments[bucketSegmentIndex];
+		final int bucketOffset = (bucket & numBucketsPerSegmentMask) << bucketSizeBits; // offset of the bucket in the segment
+		final long firstPointer = bucketSegment.getLong(bucketOffset);
+
+		try {
+			final long newFirstPointer = recordArea.appendPointerAndRecord(firstPointer, record);
+			bucketSegment.putLong(bucketOffset, newFirstPointer);
+		} catch (EOFException ex) {
+			compactOrThrow();
+			insert(record);
+			return;
+		}
+
+		numElements++;
+		resizeTableIfNecessary();
+	}
+
+	private void resizeTableIfNecessary() throws IOException {
+		if (enableResize && numElements > numBuckets) {
+			final long newNumBucketSegments = 2L * bucketSegments.length;
+			// Checks:
+			// - we can't handle more than Integer.MAX_VALUE buckets
+			// - don't take more memory than the free memory we have left
+			// - the buckets shouldn't occupy more than half of all our memory
+			if (newNumBucketSegments * numBucketsPerSegment < Integer.MAX_VALUE &&
+				newNumBucketSegments - bucketSegments.length < freeMemorySegments.size() &&
+				newNumBucketSegments < numAllMemorySegments / 2) {
+				// do the resize
+				rebuild(newNumBucketSegments);
+			}
+		}
+	}
+
+	/**
+	 * WARNING: Doing any other operation on the table invalidates the iterator! (Even
+	 * using getMatchFor of a prober!)
+	 */
+	public final class EntryIterator implements MutableObjectIterator<T> {
+
+		private final long endPosition;
+
+		public EntryIterator() {
+			endPosition = recordArea.getAppendPosition();
+			if (endPosition == 0) {
+				return;
+			}
+			recordArea.setReadPosition(0);
+		}
+
+		@Override
+		public T next(T reuse) throws IOException {
+			if (endPosition != 0 && recordArea.getReadPosition() < endPosition) {
+				// Loop until we find a non-abandoned record.
+				// Note: the last record in the record area can't be abandoned.
+				while (true) {
+					final boolean isAbandoned = recordArea.readLong() == ABANDONED_RECORD;
+					reuse = recordArea.readRecord(reuse);
```

Comment:
I implemented this (https://github.com/ggevay/flink/commit/7f789e538c475d5467705fbf154221c0b06507a8), and it brought about a 5% performance gain, when there are lots of size-changes.